### PR TITLE
Do eager tracking of moves

### PR DESCRIPTION
**Pull request by ianhi**
Fixes: https://github.com/earth-mover/icechunk/issues/1562

We weren't correctly calculating `get_node` on any non root nodes when a node had been moved into that node from elsewhere.

So there are three plcaes to fix

1. movetracker

 The simplest fix for this seems to be to eagerly compute the consequences of all moves as we accumulate them, instead of at the end. This is very useful for the case of multi move sequences

`/a -> /b/a`,  `/b->/c`. when we list what is under `/c` we need to be able to capture `a`

I think the total amount of iteration is the same or less then before because when commit is called ultimately all nodes get iterated over to track if they moved or not.

I added two data structures to MoveTracker. To track moves based on final or original key. This is because we need efficient lookups by both at different points.

2. move_node

pass the subtree info to movetracker so it can properly compute everything

3. get_node

use the new moved_into to figure out what got movedinto a group from elsewhere.

### Testing

added lots of rust tests with weird edge cases I ran into while figuring this out. I used some helpers to remove some of the visual noise of unwrap etc as I find it much easier to follow that way. but happy to revert that stuff as it is not actually necessary.

Also added hypothesis based tree move tests which is how i found all those edge cases. passes consistently now. so pretty happy with that.

### CODE REVIEW

File: icechunk-python/python/icechunk/testing/trees.py [hunk 1]
```diff
@@ -76,15 +76,64 @@ def _write(group: zarr.Group, node: GroupNode) -> None:
 
 Node = ArrayNode | GroupNode
 
+# Fixed dtype/shape — we're testing path handling, not dtype serialization.
+DTYPE = np.dtype("i4")
+SHAPE = (1,)
+
+
+def tree_from_str(spec: str) -> GroupNode:
```
Comment (ianhi) [inline icechunk-python/python/icechunk/testing/trees.py:85]: exists solely to make the `@example`s for hypothesis more human readable.

**Comment by ianhi** [comment]
there may be a bit of extra verbosity in the comments - a consequence of me slowly going through the rust and leaving (or having claude leave me) notes. believe ti or this is hte pared back version, but I can try to trim them down more
